File: src/12-B-weather_airq.R
library(tidyverse)
library(data.table)
library(arrow)
library(sf)

# Air Quality ----
POP_COM <- fread("data/external/insee-population/base-cc-evol-struct-pop-2021.csv.gz") |>
  select(code_insee24 = CODGEO, pop_tot = P21_POP)

TABLE_PASSAGE_DF <- read_feather("results_building/t_passage.feather") |>
  select(code_insee24, arr24) |>
  unique()

ARR2COM <- read_feather("results_building/arr2com.feather") |>
  left_join(TABLE_PASSAGE_DF, by=c("code_com"="code_insee24")) |>
  select(code_insee24 = code_arr, arr24)

TABLE_PASSAGE_DF <- TABLE_PASSAGE_DF |> rbind(ARR2COM)

GEOCOM <- read_sf("data/external/communes-50m.geojson") |>
  select(code_insee24 = code, geometry)

WEATHER <- fread("data/external/mfrance-meteo/meteo_2017_2024_plus.csv") |>
  st_as_sf(coords = c("LON", "LAT"), crs = 4326) 

names(WEATHER) <- names(WEATHER) |> tolower()

WEATHER_GEOMS <- WEATHER |> 
  select(geometry) |> 
  unique() |>
  st_join(GEOCOM, join = st_within) |>
  filter(!code_insee24%in%c("75056", "13055", "69123")) |>
  mutate(wkt = as.character(geometry)) |>
  st_set_geometry(NULL)

WEATHER <- WEATHER |>
  mutate(wkt = as.character(geometry)) |>
  st_set_geometry(NULL) |>
  left_join(WEATHER_GEOMS, by = "wkt") |>
  left_join(POP_COM, by="code_insee24") 

WEATHER <- WEATHER |>
  left_join(TABLE_PASSAGE_DF, by="code_insee24") |>
  mutate(
    year = substr(aaaamm, 1,4),
    month = substr(aaaamm, 5, 6)
  ) 

sum_weather <- function(DATA, geo_level, time_level){
  DATA |>
    group_by({{geo_level}}, {{time_level}}) |>
    summarise(across(c(tm, txab, tnab, umm, rr, inst), ~ weighted.mean(.x, pop_tot, na.rm = TRUE))) |>
    ungroup() 
  #NB - This approximate values for humidity for small arr near Paris using data from near Paris arr
}


WEATHER_ARR <- WEATHER |>
  mutate(
    season = case_when(
      month%in%c("12", "01", "02") ~ "winter",   
      month%in%c("03", "04", "05") ~ "spring",
      month%in%c("06", "07", "08") ~ "summer",
      month%in%c("09", "10", "11") ~ "autumn"
    )
  ) |>
  sum_weather(arr24, season) |>
  pivot_wider(names_from = season, values_from = c(tm, txab, tnab, umm, rr, inst)) |>
  select(-c(starts_with("inst"), txab_autumn, txab_spring, txab_winter, tnab_autumn, tnab_spring, tnab_summer)) |>
  mutate(across(where(is.numeric), ~ zoo::na.approx(.x, rule = 2, na.rm = FALSE))) 
  

WEATHER23_DEP <- WEATHER |>
  mutate(dep = substr(arr24, 1, 2)) |>
  filter(year == "2023") |>
  sum_weather(dep, month) |>
  filter(!is.na(tm)) |>
  pivot_wider(names_from = month, values_from = c(tm, txab, tnab, umm, rr, inst)) |>
  mutate(across(where(is.numeric), ~ zoo::na.approx(.x, rule = 2, na.rm = FALSE))) |>
  pivot_longer(-dep, names_to = c(".value", "month"), names_sep = "_")

## Exports ----
write_feather(as.data.frame(WEATHER_ARR), "data/interim/arrondissements/weather.feather", compression = "zstd")
write_feather(as.data.frame(WEATHER23_DEP), "data/interim/departement/weather_2023.feather", compression = "zstd")


# Air quality ----
TABLE_PASSAGE_DF2 <- read_feather("results_building/t_passage.feather") |>
  select(code_insee18, arr24, bv2022) |>
  unique()

AIR_QUAL <- fread("data/external/medd-qualite_air/qualite_air_clean_final.csv") |>
  mutate(code_insee18 = str_pad(CODE_INSEE, 5, pad = "0")) |> 
  inner_join(TABLE_PASSAGE_DF2, by="code_insee18")

names(AIR_QUAL) <- names(AIR_QUAL) |> tolower()

sum_airQ <- function(DATA, level){
  DATA <- DATA  |>
    group_by({{level}}) |>
    summarise(across(c(no2_mean_concentration, o3_mean_concentration, somo35_mean, aot40_mean, pm10_mean_concentration, pm25_mean_concentration), ~ weighted.mean(.x, population, na.rm = TRUE)))
}

AIR_QUAL_ARR <- AIR_QUAL |> sum_airQ(arr24)
AIR_QUAL_BV22 <- AIR_QUAL |> sum_airQ(bv2022)

## Export ----
write_feather(as.data.frame(AIR_QUAL_ARR), "data/interim/arrondissements/air_quality.feather", compression = "zstd")
write_feather(as.data.frame(AIR_QUAL_BV22), "data/interim/bv2022/air_quality.feather", compression = "zstd")

# GHG emissions ----
TABLE_PASSAGE_DF3 <- read_feather("results_building/t_passage.feather") |>
  select(code_insee21, arr24, bv2022) |>
  unique()

GES <- fread("data/external/medd-qualite_air/emission_gaz_2021_clean_final.csv") |>
  mutate(code_insee21 = str_pad(CODE_COMMUNE_2021, 5, pad= "0")) |>
  select(
    code_insee21, ges_total = EMISSIONS_GES_TOTAL, 
    ges_routier = EMISSIONS_GES_ROUTIER, 
    ges_resid = EMISSIONS_GES_RESIDENTIEL
  ) |>
  left_join(TABLE_PASSAGE_DF3, by="code_insee21") 

sum_ges <- function(DATA, level){
  DATA |>
    group_by({{level}}) |>
    summarise(across(c(ges_routier, ges_resid), ~sum(.x, na.rm = T)))
}

GES_ARR <- GES |> sum_ges(arr24)
GES_BV22 <- GES |> sum_ges(bv2022)

## Export ----
write_feather(as.data.frame(GES_ARR), "data/interim/arrondissements/greenhouse.feather", compression = "zstd")
write_feather(as.data.frame(GES_BV22), "data/interim/bv2022/greenhouse.feather", compression = "zstd")

File: src/12-B-weather_airq.py
import pandas as pd
import geopandas as gpd
import numpy as np

# --- Air Quality: Population and Passage Data ---
pop_com = pd.read_csv(
    "data/external/insee-population/base-cc-evol-struct-pop-2021.csv.gz",
    delimiter=";",
    low_memory=False,
    compression="gzip",
    usecols=["CODGEO", "P21_POP"]
).rename(columns={"CODGEO": "code_insee24", "P21_POP": "pop_tot"})

# Passage table and arrondissement-to-commune mapping
table_passage_df = (
    pd.read_feather("results_building/t_passage.feather")[["code_insee24", "arr24"]]
    .drop_duplicates()
)

arr2com = (
    pd.read_feather("results_building/arr2com.feather")
    .merge(
        table_passage_df,
        left_on="code_com",
        right_on="code_insee24",
        how="left"
    )[["code_arr", "arr24"]]
    .rename(columns={"code_arr": "code_insee24"})
)

table_passage_df = pd.concat([table_passage_df, arr2com], ignore_index=True)

# Commune geometries
geocom = gpd.read_file("data/external/communes-50m.geojson")[["code", "geometry"]]
geocom = geocom.rename(columns={"code": "code_insee24"})

# Weather Data 
weather_df = pd.read_csv(
    "data/external/mfrance-meteo/meteo_2017_2024_plus.csv"
)

weather_gdf = gpd.GeoDataFrame(
    weather_df,
    geometry=gpd.points_from_xy(weather_df.LON, weather_df.LAT),
    crs="EPSG:4326"
)
weather_gdf.columns = weather_gdf.columns.str.lower()

### Spatial join: map stations to communes
weather_geoms = (
    weather_gdf[["geometry"]]
    .drop_duplicates()
    .pipe(lambda g: gpd.sjoin(g, geocom, how="left", predicate="within"))
    .loc[lambda df: ~df["code_insee24"].isin(["75056", "13055", "69123"])]
)
weather_geoms["wkt"] = weather_geoms.geometry.apply(lambda geom: geom.wkt)
weather_geoms = weather_geoms.drop(columns="geometry")

## Flatten weather and join attributes
weather = weather_gdf.copy()
weather["wkt"] = weather.geometry.apply(lambda geom: geom.wkt)
weather = weather.drop(columns="geometry")
weather = (
    weather
    .merge(weather_geoms[["wkt", "code_insee24"]], on="wkt", how="left")
    .merge(pop_com, on="code_insee24", how="left")
    .merge(table_passage_df, on="code_insee24", how="left")
)
weather["year"] = weather.aaaamm.astype(str).str[:4]
weather["month"] = weather.aaaamm.astype(str).str[4:6]



# Weighted summary function for weather metrics
def sum_weather(df: pd.DataFrame, geo_level: str, time_level: str) -> pd.DataFrame:
    metrics = ["tm", "txab", "tnab", "umm", "rr", "inst"]
    def agg(grp):
        out = {}
        for col in metrics:
            vals = grp[col]
            w    = grp["pop_tot"]
            mask = vals.notna() & w.notna()
            if mask.any():
                out[col] = np.average(vals[mask], weights=w[mask])
            else:
                out[col] = np.nan
        return pd.Series(out)
    return (
        df
        .groupby([geo_level, time_level])
        .apply(agg)
        .reset_index()
    )


## ------- Arrondissement-level seasonal summary ---------
weather["season"] = np.select(
    [weather.month.isin(["12","01","02"]),
     weather.month.isin(["03","04","05"]),
     weather.month.isin(["06","07","08"]),
     weather.month.isin(["09","10","11"])]
    , ["winter", "spring", "summer", "autumn"],
    default=None
)
weather_arr = sum_weather(weather, "arr24", "season")
weather_arr = (
    weather_arr
    .pivot(
        index="arr24", columns="season",
        values=["tm","txab","tnab","umm","rr","inst"]
    )
    .reset_index()
)

weather_arr.columns = ["_".join(filter(None, col)).strip("_") for col in weather_arr.columns.values]

drop_cols = [c for c in weather_arr if c.startswith("inst_")] + [
    "txab_autumn", "txab_spring", "txab_winter",
    "tnab_autumn", "tnab_spring", "tnab_summer"
]

weather_arr = weather_arr.drop(columns=drop_cols)
num_cols = weather_arr.select_dtypes(include=np.number).columns
weather_arr[num_cols] = weather_arr[num_cols].interpolate(
    method="linear", axis=0, limit_direction="both"
)

#------------ Department level monthly summary --------------
weather23_dep = (
    weather
      .assign(dep=weather.arr24.str[:2])
      .loc[lambda df: df.year == "2023"]
      .pipe(lambda df: sum_weather(df, "dep", "month"))
).dropna(subset=["tm"])

weather23_dep = weather23_dep.pivot(
    index="dep",
    columns="month",
    values=["tm","txab","tnab","umm","rr","inst"]
)

weather23_dep.columns = [
    f"{var}_{month}"
    for var, month in weather23_dep.columns
]

num_cols = weather23_dep.select_dtypes(include="number").columns
for col in num_cols:
    weather23_dep[col] = weather23_dep[col].interpolate(
        method="linear",
        axis=0,
        limit_direction="both"
    )

weather23_dep = (
    weather23_dep
      .reset_index()
      .melt(
         id_vars="dep",
         var_name="var_month",
         value_name="value"
      )
      .assign(
         var   = lambda df: df.var_month.str.split("_").str[0],
         month = lambda df: df.var_month.str.split("_").str[1]
      )
      .pivot(
         index=["dep","month"],
         columns="var",
         values="value"
      )
      .reset_index()
)

#-------- Export results to Feather ----------
weather_arr.to_feather(
    "data/interim/arrondissements/weather.feather",
    compression="zstd"
)
weather23_dep.to_feather(
    "data/interim/departement/weather_2023.feather",
    compression="zstd"
)

# --------- Air Quality Aggregation -------------

table_passage_df2 = (
    pd.read_feather("results_building/t_passage.feather")
    [["code_insee18", "arr24", "bv2022"]]
    .drop_duplicates()
)
air_qual = pd.read_csv("data/external/medd-qualite_air/qualite_air_clean_final.csv")
air_qual["code_insee18"] = (
    air_qual.CODE_INSEE.astype(str).str.zfill(5)
)
air_qual = air_qual.merge(table_passage_df2, on="code_insee18", how="inner")
air_qual.columns = air_qual.columns.str.lower()

def sum_airQ(df: pd.DataFrame, level: str) -> pd.DataFrame:
    metrics = [
        "no2_mean_concentration", "o3_mean_concentration", "somo35_mean",
        "aot40_mean", "pm10_mean_concentration", "pm25_mean_concentration"
    ]
    def agg(grp):
        out = {}
        for col in metrics:
            vals = grp[col]
            w    = grp["population"]
            mask = vals.notna() & w.notna()
            if mask.any():
                out[col] = np.average(vals[mask], weights=w[mask])
            else:
                out[col] = np.nan
        return pd.Series(out)
    return (
        df
        .groupby(level)
        .apply(agg)
        .reset_index()
    )

air_qual_arr = sum_airQ(air_qual, "arr24")
air_qual_bv22 = sum_airQ(air_qual, "bv2022")

air_qual_arr.to_feather(
    "data/interim/arrondissements/air_quality.feather",
    compression="zstd"
)
air_qual_bv22.to_feather(
    "data/interim/bv2022/air_quality.feather",
    compression="zstd"
)


# ---------- Greenhouse Gas Emissions Aggregation ----------------
table_passage_df3 = (
    pd.read_feather("results_building/t_passage.feather")
    [["code_insee21", "arr24", "bv2022"]]
    .drop_duplicates()
)

ges = pd.read_csv("data/external/medd-qualite_air/emission_gaz_2021_clean_final.csv")

ges["code_insee21"] = ges.CODE_COMMUNE_2021.astype(str).str.zfill(5)

ges = ges[[
    "code_insee21", "EMISSIONS_GES_TOTAL",
    "EMISSIONS_GES_ROUTIER", "EMISSIONS_GES_RESIDENTIEL"
]].rename(
    columns={
        "EMISSIONS_GES_ROUTIER": "ges_routier",
        "EMISSIONS_GES_RESIDENTIEL": "ges_resid"
    }
)
ges = ges.merge(table_passage_df3, on="code_insee21", how="left")

def sum_ges(df: pd.DataFrame, level: str) -> pd.DataFrame:
    metrics = ["ges_routier", "ges_resid"]
    return df.groupby(level)[metrics].sum().reset_index()

ges_arr = sum_ges(ges, "arr24")
ges_bv22 = sum_ges(ges, "bv2022")

ges_arr.to_feather(
    "data/interim/arrondissements/greenhouse.feather",
    compression="zstd"
)
ges_bv22.to_feather(
    "data/interim/bv2022/greenhouse.feather",
    compression="zstd"
)
